heron/common/src/python/argparser.py

```diff
@@ -0,0 +1,196 @@
+
+          if command in cls.cmdmap[app] and env in cls.cmdmap[app][command]:
+            cls.cmdmap[app][command][env] = cls.cmdmap[app][command][env] + ' ' + args_list_string
+          else:
+            cls.cmdmap[app][command][env] = args_list_string
+      Log.debug("RC cmdmap %s", json.dumps(cls.cmdmap))
+    else:
+      Log.warn("%s is not an existing file", effective_rc)
+
+  # for each command / cluster-role-env combination, get the commands from heronrc
+  # remove any duplicates that have already been supplied already  and
+  # present in the command-dictionary
+  @classmethod
+  def get_args_for_command_role(cls, app, command, role):
+    args_for_command_role = ''
+    if app in cls.cmdmap and command in cls.cmdmap[app] and role in cls.cmdmap[app][command]:
+      args_for_command_role = (cls.cmdmap[app][command][role],
+                               args_for_command_role)[cls.cmdmap[app][command][role] is None]
+    return args_for_command_role.split()
+
+  # this is invoked when the parser.parse_args is called
+  #  apply the commands in the following precedence order
+  #  use the defaults in the command line
+
+  def _read_args_from_files(self, arg_strings):
+    new_arg_strings = []
+    command = self.rccommand
+    if len(sys.argv) > 1:
+      command = (sys.argv[1], self.rccommand)[self.rccommand != '']
+    role = self.rcclusterrole
+    if len(sys.argv) > 2:
+      role = (sys.argv[2], self.rcclusterrole)[self.rcclusterrole != '']
+    app = self.prog
+    new_arg_strings.extend(
+        self.get_args_for_command_role(app, command, role))
+    new_arg_strings.extend(
+        self.get_args_for_command_role(app, command, '*'))
+    new_arg_strings.extend(self.get_args_for_command_role(app, '*', '*'))
+    new_arg_strings.extend(self.get_args_for_command_role('*', '*', '*'))
+    arg_strings.extend(new_arg_strings)
+    return arg_strings
+
+  def parse_known_args(self, args=None, namespace=None):
+    namespace, args = super(HeronRCArgumentParser,
+                            self).parse_known_args(args, namespace)
+    dict_ns = namespace.__dict__
+    if self.prog == 'heron':
+      try:
+        for key in dict_ns:
+          val = dict_ns[key]
+          if val is not None and isinstance(val, list) and len(val) > 0:
+            dict_ns[key] = val[0]
+      except Exception:
+        Log.warn("heronrc: unable to clobber arguments (%s,%s ) ", namespace, args)
+        Log.debug(traceback.format_exc())
+    return namespace, args
+
+def main():
+  parser = HeronRCArgumentParser(
+      prog='heron',
+      epilog=help_epilog,
+      formatter_class=config.SubcommandHelpFormatter,
+      fromfile_prefix_chars='@',
+      add_help=False,
+      rcfile="./.heronrc")
+  parser.add_subparsers(
+      title="Available commands",
+      metavar='<command> <options>')
+
+  args, unknown_args = parser.parse_known_args()
+  Log.info("parse results args: %s  unknown: %s ", args, unknown_args)
```

Comment:
this is the main stub. meant to print it to stdout. info seems okay to me. let me know if you feel strongly